src/boosting/gbdt.cpp

```diff
@@ -58,6 +63,12 @@ void GBDT::Init(const Config* config, const Dataset* train_data, const Objective
   es_first_metric_only_ = config_->first_metric_only;
   shrinkage_rate_ = config_->learning_rate;
 
+#ifdef USE_CUDA
```

Comment:
This is needed.

Reply:
This is needed.

Reply:
Could you please elaborate why is it needed?

Reply:
current_device and current_learning are CUDA specific.  They were designed to allow the CUDA compiled code to be run in either CUDA or CPU mode without recompilation.  My group's job is just to get this PR released and not to majorly rewrite the code.  There is another group working on some future features enhancements but we are hoping to have the OSC take over a majority of the workload.

Reply:
> They were designed to allow the CUDA compiled code to be run in either CUDA or CPU mode without recompilation

OpenCL implementation can be run in GPU and CPU mode as well.

Default values of `current_device` and `current_learning` from `include/LightGBM/cuda/vector_cudahost.h` are set to CPU. So I still believe there are a lot of excess `#ifdef USE_CUDA` guards that can be omitted.

> My group's job is just to get this PR released and not to majorly rewrite the code. There is another group working on some future features enhancements but we are hoping to have the OSC take over a majority of the workload.

Unfortunately, we are not aware of your internal company structure. We just see a PR that should be reviewed to be included into the project. Each PR should bring some value to the project, improve it in any way and which is very important be complete or at least we should have some **public** TODO list (refer for example to https://github.com/microsoft/LightGBM/pull/2305#issuecomment-574560469).

I believe it will be better to invite that group who is responsible for the code.

Indeed, open source community is extremely powerful thing! But unfortunately, according to our experience, there are not so many experts in GPU computing visiting our repo, especially those who are using POWER architecture (as this implementation is proven to be working on ppc64le).

Reply:
I have done minimum testing on x64 with Ubuntu and CUDA.  Last time I tested, all unit tests work and pass.

Reply:
This is really good news! We will setup Ubuntu with GPU hardware in our CI pool with manual triggering once this PR will be merged (https://github.com/microsoft/LightGBM/pull/3160#issuecomment-659422651).